File: clients/go/examples/measure_latency.go
package main

import (
	"fmt"
	"log"
	"sort"
	"time"
	blazecache "github.com/blazecache/client"
)

func main() {
	serverAddr := "127.0.0.1:6792"
	
	fmt.Printf("=== Network Latency Measurement ===\n")
	fmt.Printf("Server: %s\n\n", serverAddr)
	
	// Create client
	client, err := blazecache.New(serverAddr)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	
	// Test ping first
	if err := client.Ping(); err != nil {
		log.Fatalf("Ping failed: %v", err)
	}
	fmt.Println("✓ Server connection verified\n")
	
	// Measure PING latency
	fmt.Println("=== PING Latency ===")
	pingLatencies := make([]time.Duration, 100)
	for i := 0; i < 100; i++ {
		start := time.Now()
		if err := client.Ping(); err != nil {
			log.Fatalf("Ping failed: %v", err)
		}
		pingLatencies[i] = time.Since(start)
	}
	printStats("PING", pingLatencies)
	
	// Measure SET latency
	fmt.Println("\n=== SET Latency ===")
	setLatencies := make([]time.Duration, 100)
	for i := 0; i < 100; i++ {
		key := fmt.Sprintf("latency-test-key-%d", i)
		value := []byte(fmt.Sprintf("value-%d", i))
		item := &blazecache.Item{Key: key, Value: value}
		
		start := time.Now()
		if err := client.Set(item); err != nil {
			log.Fatalf("Set failed: %v", err)
		}
		setLatencies[i] = time.Since(start)
	}
	printStats("SET", setLatencies)
	
	// Measure GET latency (warm cache)
	fmt.Println("\n=== GET Latency (cache hit) ===")
	getLatencies := make([]time.Duration, 100)
	for i := 0; i < 100; i++ {
		key := fmt.Sprintf("latency-test-key-%d", i)
		
		start := time.Now()
		result, err := client.Get(key)
		if err != nil {
			log.Fatalf("Get failed: %v", err)
		}
		if result == nil {
			log.Fatalf("Get returned nil")
		}
		getLatencies[i] = time.Since(start)
		_ = result
	}
	printStats("GET (hit)", getLatencies)
	
	// Measure GET latency (cache miss)
	fmt.Println("\n=== GET Latency (cache miss) ===")
	getMissLatencies := make([]time.Duration, 100)
	for i := 0; i < 100; i++ {
		key := fmt.Sprintf("latency-test-miss-%d", i)
		
		start := time.Now()
		_, err := client.Get(key)
		if err == nil {
			log.Fatalf("Expected error for missing key")
		}
		getMissLatencies[i] = time.Since(start)
	}
	printStats("GET (miss)", getMissLatencies)
	
	// Measure SET+GET combined latency
	fmt.Println("\n=== SET+GET Combined Latency ===")
	combinedLatencies := make([]time.Duration, 100)
	for i := 0; i < 100; i++ {
		key := fmt.Sprintf("latency-test-combined-%d", i)
		value := []byte(fmt.Sprintf("value-%d", i))
		item := &blazecache.Item{Key: key, Value: value}
		
		start := time.Now()
		if err := client.Set(item); err != nil {
			log.Fatalf("Set failed: %v", err)
		}
		result, err := client.Get(key)
		if err != nil {
			log.Fatalf("Get failed: %v", err)
		}
		if result == nil {
			log.Fatalf("Get returned nil")
		}
		combinedLatencies[i] = time.Since(start)
		_ = result
	}
	printStats("SET+GET", combinedLatencies)
	
	fmt.Println("\n=== Summary ===")
	fmt.Printf("PING RTT:     %s\n", formatDuration(median(pingLatencies)))
	fmt.Printf("SET RTT:      %s\n", formatDuration(median(setLatencies)))
	fmt.Printf("GET (hit) RTT: %s\n", formatDuration(median(getLatencies)))
	fmt.Printf("GET (miss) RTT: %s\n", formatDuration(median(getMissLatencies)))
	fmt.Printf("SET+GET RTT:  %s\n", formatDuration(median(combinedLatencies)))
}

func printStats(name string, latencies []time.Duration) {
	sorted := make([]time.Duration, len(latencies))
	copy(sorted, latencies)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i] < sorted[j]
	})
	
	min := sorted[0]
	max := sorted[len(sorted)-1]
	median := sorted[len(sorted)/2]
	p95 := sorted[int(float64(len(sorted))*0.95)]
	p99 := sorted[int(float64(len(sorted))*0.99)]
	
	var sum time.Duration
	for _, l := range latencies {
		sum += l
	}
	avg := sum / time.Duration(len(latencies))
	
	fmt.Printf("  Min:    %s\n", formatDuration(min))
	fmt.Printf("  Max:    %s\n", formatDuration(max))
	fmt.Printf("  Avg:    %s\n", formatDuration(avg))
	fmt.Printf("  Median: %s\n", formatDuration(median))
	fmt.Printf("  P95:    %s\n", formatDuration(p95))
	fmt.Printf("  P99:    %s\n", formatDuration(p99))
}

func median(durations []time.Duration) time.Duration {
	sorted := make([]time.Duration, len(durations))
	copy(sorted, durations)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i] < sorted[j]
	})
	return sorted[len(sorted)/2]
}

func formatDuration(d time.Duration) string {
	if d < time.Microsecond {
		return fmt.Sprintf("%.2fns", float64(d.Nanoseconds()))
	} else if d < time.Millisecond {
		return fmt.Sprintf("%.2fµs", float64(d.Nanoseconds())/1000.0)
	} else {
		return fmt.Sprintf("%.2fms", float64(d.Nanoseconds())/1000000.0)
	}
}

File: clients/rust/examples/measure_latency.rs
use blazecache_client::Client;
use std::time::{Duration, Instant};
use std::collections::VecDeque;

fn format_duration(d: Duration) -> String {
    let nanos = d.as_nanos() as f64;
    if nanos < 1000.0 {
        format!("{:.2}ns", nanos)
    } else if nanos < 1_000_000.0 {
        format!("{:.2}µs", nanos / 1000.0)
    } else {
        format!("{:.2}ms", nanos / 1_000_000.0)
    }
}

fn print_stats(name: &str, latencies: &[Duration]) {
    let mut sorted = latencies.to_vec();
    sorted.sort();
    
    let min = sorted[0];
    let max = sorted[sorted.len() - 1];
    let median = sorted[sorted.len() / 2];
    let p95_idx = (sorted.len() as f64 * 0.95) as usize;
    let p95 = sorted[p95_idx.min(sorted.len() - 1)];
    let p99_idx = (sorted.len() as f64 * 0.99) as usize;
    let p99 = sorted[p99_idx.min(sorted.len() - 1)];
    
    let sum: Duration = latencies.iter().sum();
    let avg = sum / latencies.len() as u32;
    
    println!("  Min:    {}", format_duration(min));
    println!("  Max:    {}", format_duration(max));
    println!("  Avg:    {}", format_duration(avg));
    println!("  Median: {}", format_duration(median));
    println!("  P95:    {}", format_duration(p95));
    println!("  P99:    {}", format_duration(p99));
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let server_addr = "127.0.0.1:6792";
    
    println!("=== Network Latency Measurement ===");
    println!("Server: {}\n", server_addr);
    
    let client = Client::new(vec![server_addr.to_string()]).await?;
    
    // Test ping first
    client.ping().await?;
    println!("✓ Server connection verified\n");
    
    // Measure PING latency
    println!("=== PING Latency ===");
    let mut ping_latencies = Vec::new();
    for i in 0..100 {
        let start = Instant::now();
        client.ping().await?;
        ping_latencies.push(start.elapsed());
    }
    print_stats("PING", &ping_latencies);
    
    // Measure SET latency
    println!("\n=== SET Latency ===");
    let mut set_latencies = Vec::new();
    for i in 0..100 {
        let key = format!("latency-test-key-{}", i);
        let value = format!("value-{}", i).into_bytes();
        
        let start = Instant::now();
        client.set(&key, &value, 3600).await?;
        set_latencies.push(start.elapsed());
    }
    print_stats("SET", &set_latencies);
    
    // Measure GET latency (warm cache)
    println!("\n=== GET Latency (cache hit) ===");
    let mut get_latencies = Vec::new();
    for i in 0..100 {
        let key = format!("latency-test-key-{}", i);
        
        let start = Instant::now();
        let _result = client.get(&key).await?;
        get_latencies.push(start.elapsed());
    }
    print_stats("GET (hit)", &get_latencies);
    
    // Measure SET+GET combined latency
    println!("\n=== SET+GET Combined Latency ===");
    let mut combined_latencies = Vec::new();
    for i in 0..100 {
        let key = format!("latency-test-combined-{}", i);
        let value = format!("value-{}", i).into_bytes();
        
        let start = Instant::now();
        client.set(&key, &value, 3600).await?;
        let _result = client.get(&key).await?;
        combined_latencies.push(start.elapsed());
    }
    print_stats("SET+GET", &combined_latencies);
    
    println!("\n=== Summary ===");
    println!("PING RTT:     {}", format_duration(ping_latencies[ping_latencies.len() / 2]));
    println!("SET RTT:      {}", format_duration(set_latencies[set_latencies.len() / 2]));
    println!("GET (hit) RTT: {}", format_duration(get_latencies[get_latencies.len() / 2]));
    println!("SET+GET RTT:  {}", format_duration(combined_latencies[combined_latencies.len() / 2]));
    
    Ok(())
}

